### PR TITLE
Partial srg compilation + Metal rendering fixes

### PR DESCRIPTION
**Pull request by moudgils**
- Partial SRG compilation support where a SRG will not update a resource type if it wasn't updated
  - SharedResourceGroupData has a update mask which is updated by RPI when SRGdata is updated
  - This mask is then copied over to RHI at which point RHI keeps the bit alive for 3 compiles as SRGs are triple buffered.
  - The RHI backend uses the bits within the update mask to figure out when to compile the appropriate type.
  - This way higher level code set views once and can call compile each frame without paying the penalty for the view
  - Invalidated views are also tracked to ensure they are updated appropriately. This is done via tracking the hash associated with the views. If the hash changes then RHI will set the appropriate bit within the update mask.
  - Added a cvar to disable partial SRG compilation in order to eliminate this as a potential cause for a bug in future

- Metal fixes to ensure the cpu is not stomping over gpu memory that is not consumed yet.
- Misc cleanup
- Tested ASV full test suite using dx12 and Vk

### CODE REVIEW

**Comment by moudgils** [comment]
All the tests passed. The red lines are due to the following Trace:error which seems harmless
Automation: Trace::Error
 C:/GitHub/o3de/AtomSampleViewer/Gem/Code/Source/Automation/ScriptManager.cpp(320): 'void __cdecl AtomSampleViewer::ScriptManager::TickScript(float)'


AtomSampleViewerStandalone_KAtMv7jzWs

**Comment by moudgils** [comment]
3qXRHK4EZl

File: Gems/Atom/RHI/Code/Source/RHI/ShaderResourceGroupPool.cpp [hunk 1]
```diff
@@ -246,6 +250,113 @@ namespace AZ
...
+        template<typename T>
+        void ShaderResourceGroupPool::UpdateMaskBasedOnViewHash(
+            ShaderResourceGroup& shaderResourceGroup, Name entryName, AZStd::span<const RHI::ConstPtr<T>> views,
+            ShaderResourceGroupData::ResourceType resourceType)
+        {
+            //Get the view hash and check if it was updated in which case we need to compile those views. 
+            HashValue64 viewHash = GetViewHash<T>(views);
+            if (shaderResourceGroup.GetViewHash(entryName) != viewHash)
+            {
+                shaderResourceGroup.EnableRhiResourceTypeCompilation(static_cast<ShaderResourceGroupData::ResourceTypeMask>(AZ_BIT(static_cast<uint32_t>(resourceType))));
```
Comment (rgba16f) [inline Gems/Atom/RHI/Code/Source/RHI/ShaderResourceGroupPool.cpp:276]: nit: line length. Perhaps we could get a utility function to convert a resource type to a ResourceTypeMask.

**Review by rgba16f — APPROVED** [review]

**Review by dmcdiarmid-ly — APPROVED** [review]

File: Gems/Atom/RHI/DX12/Code/Source/RHI/ImageView.cpp [hunk 2]
```diff
@@ -61,6 +61,9 @@ namespace AZ
                     m_depthStencilReadDescriptor);
             }
 
+            m_hash = TypeHash64(viewDescriptor.GetHash(), m_hash);
+            m_hash = TypeHash64(m_format, m_hash);
```
Comment (jiaweig-amzn) [inline Gems/Atom/RHI/DX12/Code/Source/RHI/ImageView.cpp:65]: Besides format, bindFlags also has a source from the image, but it is not cached in ImageView. Should we take this into consideration and maybe has with Image::GetHash as well?

**Review by jiaweig-amzn — APPROVED** [review]

**Pull request merged**